File: tyhp-lang/phpstorm/src/main/kotlin/com/tyhp/lang/workspace/SelectOwner.kt
package com.tyhp.lang.workspace

interface OwnerCandidate {
    val projectFilePath: String
    val projectDir: String
}

data class SimpleOwner(
    override val projectFilePath: String,
    override val projectDir: String,
) : OwnerCandidate

private fun comparePathThenLex(a: OwnerCandidate, b: OwnerCandidate): Int {
    val byLength = a.projectFilePath.length.compareTo(b.projectFilePath.length)
    if (byLength != 0) {
        return byLength
    }
    return a.projectFilePath.compareTo(b.projectFilePath)
}

private fun pickBest(candidates: List<OwnerCandidate>): OwnerCandidate =
    candidates.minWith(Comparator(::comparePathThenLex))

private fun nearestAncestor(
    filePath: String,
    candidates: List<OwnerCandidate>,
    caseInsensitive: Boolean,
): OwnerCandidate? {
    val ancestors = candidates.filter { isPathInside(it.projectDir, filePath, caseInsensitive) }
    if (ancestors.isEmpty()) {
        return null
    }
    val longest = ancestors.maxOf { posixNormalize(it.projectDir).length }
    val tied = ancestors.filter { posixNormalize(it.projectDir).length == longest }
    return if (tied.size == 1) tied[0] else pickBest(tied)
}

fun selectOwner(
    filePath: String,
    candidates: List<OwnerCandidate>,
    caseInsensitive: Boolean = false,
): OwnerCandidate? {
    if (candidates.isEmpty()) {
        return null
    }
    if (candidates.size == 1) {
        return candidates[0]
    }
    nearestAncestor(filePath, candidates, caseInsensitive)?.let { return it }

    var minHops = Int.MAX_VALUE
    val nearest = ArrayList<OwnerCandidate>()
    for (candidate in candidates) {
        val hops = pathHops(candidate.projectDir, filePath, caseInsensitive)
        if (hops < minHops) {
            minHops = hops
            nearest.clear()
            nearest.add(candidate)
        } else if (hops == minHops) {
            nearest.add(candidate)
        }
    }
    return if (nearest.size == 1) nearest[0] else pickBest(nearest)
}

File: tyhp-lang/vscode/src/workspace/selectOwner.ts
/**
 * Single-owner selection when more than one `tyhp.json` include-matches a file.
 * Do not merge two language servers (they are different type worlds).
 *
 * Tie-break:
 * 1. Matching `tyhp.json` that is the nearest ancestor of the file
 * 2. Else nearest `tyhp.json` by path hops
 * 3. Else shortest `tyhp.json` path, then lexicographic
 */

import { isPathInside, pathHops, posixNormalize } from "./pathUtils";

export interface OwnerCandidate {
    readonly projectFilePath: string;
    readonly projectDir: string;
}

function comparePathThenLex(a: OwnerCandidate, b: OwnerCandidate): number {
    const la = a.projectFilePath.length;
    const lb = b.projectFilePath.length;
    if (la !== lb) {
        return la - lb;
    }
    if (a.projectFilePath < b.projectFilePath) {
        return -1;
    }
    if (a.projectFilePath > b.projectFilePath) {
        return 1;
    }
    return 0;
}

function pickBest(candidates: OwnerCandidate[]): OwnerCandidate {
    return [...candidates].sort(comparePathThenLex)[0];
}

function nearestAncestor(filePath: string, candidates: readonly OwnerCandidate[], caseInsensitive: boolean): OwnerCandidate | undefined {
    const ancestors = candidates.filter((c) => isPathInside(c.projectDir, filePath, caseInsensitive));
    if (ancestors.length === 0) {
        return undefined;
    }
    let best = ancestors[0];
    let bestLen = posixNormalize(best.projectDir).length;
    const tied: OwnerCandidate[] = [best];
    for (let i = 1; i < ancestors.length; i++) {
        const cur = ancestors[i];
        const len = posixNormalize(cur.projectDir).length;
        if (len > bestLen) {
            best = cur;
            bestLen = len;
            tied.length = 0;
            tied.push(cur);
        } else if (len === bestLen) {
            tied.push(cur);
        }
    }
    return tied.length === 1 ? tied[0] : pickBest(tied);
}

export function selectOwner(
    filePath: string,
    candidates: readonly OwnerCandidate[],
    caseInsensitive = false
): OwnerCandidate | undefined {
    if (candidates.length === 0) {
        return undefined;
    }
    if (candidates.length === 1) {
        return candidates[0];
    }

    const ancestor = nearestAncestor(filePath, candidates, caseInsensitive);
    if (ancestor) {
        return ancestor;
    }

    let minHops = Number.POSITIVE_INFINITY;
    const nearest: OwnerCandidate[] = [];
    for (const candidate of candidates) {
        const hops = pathHops(candidate.projectDir, filePath, caseInsensitive);
        if (hops < minHops) {
            minHops = hops;
            nearest.length = 0;
            nearest.push(candidate);
        } else if (hops === minHops) {
            nearest.push(candidate);
        }
    }
    return nearest.length === 1 ? nearest[0] : pickBest(nearest);
}
